File: ForexConnectAPIRobot/samples/java/WholeSample/src/com/fxcore2/samples/TradingSettingsProviderSample.java
package com.fxcore2.samples;

import com.fxcore2.O2GAccountRow;
import com.fxcore2.O2GAccountsTableResponseReader;
import com.fxcore2.O2GLoginRules;
import com.fxcore2.O2GMargin;
import com.fxcore2.O2GMarketStatus;
import com.fxcore2.O2GOfferRow;
import com.fxcore2.O2GOffersTableResponseReader;
import com.fxcore2.O2GResponse;
import com.fxcore2.O2GResponseReaderFactory;
import com.fxcore2.O2GSession;
import com.fxcore2.O2GTableType;
import com.fxcore2.O2GTradingSettingsProvider;

public class TradingSettingsProviderSample {
    private O2GSession mSession;

    private SimpleLog mSimpleLog;

    public TradingSettingsProviderSample(SimpleLog simpleLog, O2GSession session)
    {
        mSimpleLog = simpleLog;
        mSession = session;
    }

    public void print()
    {
        O2GLoginRules loginRules = mSession.getLoginRules();
        O2GResponse accountsResponse = loginRules.getTableRefreshResponse(O2GTableType.ACCOUNTS);
        if (accountsResponse == null)
            return;

        O2GResponse offersResponse = loginRules.getTableRefreshResponse(O2GTableType.OFFERS);
        if (offersResponse == null)
            return;

        O2GTradingSettingsProvider tradingSettingsProvider = loginRules.getTradingSettingsProvider();
        O2GResponseReaderFactory factory = mSession.getResponseReaderFactory();
        if (factory == null)
            return;
        
        O2GAccountsTableResponseReader accountsReader = factory.createAccountsTableReader(accountsResponse);
        if (accountsReader == null)
            return;

        O2GOffersTableResponseReader instrumentsReader = factory.createOffersTableReader(offersResponse);
        if (instrumentsReader == null)
            return;

        O2GAccountRow sAccount = accountsReader.getRow(0);
        if (sAccount == null)
            return;

        for (int i = 0; i < instrumentsReader.size(); i++)
        {
            O2GOfferRow instrumentRow = instrumentsReader.getRow(i);
            String sInstrument = instrumentRow.getInstrument();
            int iCondDistStopForTrade = tradingSettingsProvider.getCondDistStopForTrade(sInstrument);
            int iCondDistLimitForTrade = tradingSettingsProvider.getCondDistLimitForTrade(sInstrument);
            int iCondDistStopForEntryOrder = tradingSettingsProvider.getCondDistStopForEntryOrder(sInstrument);
            int iCondDistLimitForEntryOrder = tradingSettingsProvider.getCondDistLimitForEntryOrder(sInstrument);
            int iCondDistEntryStop = tradingSettingsProvider.getCondDistEntryStop(sInstrument);
            int iCondDistEntryLimit = tradingSettingsProvider.getCondDistEntryLimit(sInstrument);
            int iMinQuantity = tradingSettingsProvider.getMinQuantity(sInstrument, sAccount);
            int iMaxQuantity = tradingSettingsProvider.getMaxQuantity(sInstrument, sAccount);
            int iBaseUnitSize = tradingSettingsProvider.getBaseUnitSize(sInstrument, sAccount);
            O2GMarketStatus eMarketStatus = tradingSettingsProvider.getMarketStatus(sInstrument);
            int iMinTrailingStepForStop = tradingSettingsProvider.getMinTrailingStepForStop();
            int iMinTrailingStepForLimit = tradingSettingsProvider.getMinTrailingStepForLimit();
            int iMinTrailingStepForEntryStop = tradingSettingsProvider.getMinTrailingStepForEntryStop();
            int iMinTrailingStepForEntryLimit = tradingSettingsProvider.getMinTrailingStepForEntryLimit();
            int iMaxTrailingStepForStop = tradingSettingsProvider.getMaxTrailingStepForStop();
            int iMaxTrailingStepForLimit = tradingSettingsProvider.getMaxTrailingStepForLimit();
            int iMaxTrailingStepForEntryStop = tradingSettingsProvider.getMaxTrailingStepForEntryStop();
            int iMaxTrailingStepForEntryLimit = tradingSettingsProvider.getMaxTrailingStepForEntryLimit();
            double dMMR = tradingSettingsProvider.getMMR(sInstrument, sAccount);

            O2GMargin margin = tradingSettingsProvider.getMargins(sInstrument, sAccount);
            double dMMR2 = margin.getMMR();
            double dEMR = margin.getEMR();
            double dLMR = margin.getLMR();
            boolean b3LevelMarginUsed = margin.is3LevelMargin(); 

            mSimpleLog.log("  [{0} ({1})] Cond.Dist: ST={2}; LT={3}; SE={4}; LE={5}", sInstrument, eMarketStatus,
                iCondDistStopForTrade, iCondDistLimitForTrade, iCondDistStopForEntryOrder, iCondDistLimitForEntryOrder);
            mSimpleLog.log("            Cond.Dist entry stop={0}; entry limit={1}", iCondDistEntryStop, iCondDistEntryLimit);
            mSimpleLog.log("            Quantity: Min={0}; Max={1}. Base unit size={2}; MMR={3}", iMinQuantity,
                iMaxQuantity, iBaseUnitSize, dMMR);
            if (b3LevelMarginUsed)
                mSimpleLog.log("            Three level margin: MMR={0}; EMR={1}; LMR={2}", dMMR2, dEMR, dLMR);
            else
                mSimpleLog.log("            Single level margin: MMR={0}; EMR={1}; LMR={2}", dMMR2, dEMR, dLMR);
            mSimpleLog.log("            Margins: Min={0}; Max={1}. Base unit size={2}; MMR={3}", iMinQuantity,
                iMaxQuantity, iBaseUnitSize, dMMR);
            mSimpleLog.log("            Trailing stop: Stop={0}.{1}; Limit={2}.{3}", iMinTrailingStepForStop,
                iMaxTrailingStepForStop, iMinTrailingStepForLimit, iMaxTrailingStepForLimit);
            mSimpleLog.log("            Trailing stop: Entry Stop={0}.{1}; Entry Limit={2}.{3}",
                iMinTrailingStepForEntryStop, iMaxTrailingStepForEntryStop, iMinTrailingStepForEntryLimit, iMaxTrailingStepForEntryLimit);
        }
    }
}

File: _DO_NOT_USE_forexConnectAPIRobot/forexConnectRobot/source/TradingSettingsProviderSample.cpp
#include "stdafx.h"
#include "SimpleLog.h"
#include "TradingSettingsProviderSample.h"

/** Constructor.*/
TradingSettingsProviderSample::TradingSettingsProviderSample(SimpleLog *log, IO2GSession *session)
{
    mLog = log;
    mSession = session;
    mSession->addRef();
}

/** Destructor. */
TradingSettingsProviderSample::~TradingSettingsProviderSample()
{
}

/** Prints settings. */
void TradingSettingsProviderSample::printSettings()
{
    O2G2Ptr<IO2GLoginRules> loginRules = mSession->getLoginRules();
    O2G2Ptr<IO2GResponse> accountsResponse = loginRules->getTableRefreshResponse(Accounts);
    if (!accountsResponse)
        return;

    O2G2Ptr<IO2GResponse> offersResponse = loginRules->getTableRefreshResponse(Offers);
    if (!offersResponse)
        return;

    O2G2Ptr<IO2GTradingSettingsProvider> tradingSettingsProvider = loginRules->getTradingSettingsProvider();
    O2G2Ptr<IO2GResponseReaderFactory> factory = mSession->getResponseReaderFactory();
    if (!factory)
        return;
    O2G2Ptr<IO2GAccountsTableResponseReader> accountsReader = factory->createAccountsTableReader(accountsResponse);
    if (!accountsReader)
        return;

    O2G2Ptr<IO2GOffersTableResponseReader> instrumentsReader = factory->createOffersTableReader(offersResponse);
    if (!instrumentsReader)
        return;

    O2G2Ptr<IO2GAccountRow> account = accountsReader->getRow(0);
    if (!account)
        return;
    for (int i = 0; i < instrumentsReader->size(); ++i)
    {
        O2G2Ptr<IO2GOfferRow> instrumentRow = instrumentsReader->getRow(i);
        const char* instrument = instrumentRow->getInstrument();
        int condDistStopForTrade = tradingSettingsProvider->getCondDistStopForTrade(instrument);
        int condDistLimitForTrade = tradingSettingsProvider->getCondDistLimitForTrade(instrument);
        int condDistStopForEntryOrder = tradingSettingsProvider->getCondDistStopForEntryOrder(instrument);
        int condDistLimitForEntryOrder = tradingSettingsProvider->getCondDistLimitForEntryOrder(instrument);
        int condDistEntryStop = tradingSettingsProvider->getCondDistEntryStop(instrument);
        int condDistEntryLimit = tradingSettingsProvider->getCondDistEntryLimit(instrument);
        int minQuantity = tradingSettingsProvider->getMinQuantity(instrument, account);
        int maxQuantity = tradingSettingsProvider->getMaxQuantity(instrument, account);
        int baseUnitSize = tradingSettingsProvider->getBaseUnitSize(instrument, account);
        O2GMarketStatus marketStatus = tradingSettingsProvider->getMarketStatus(instrument);
        int minTrailingStepForStop = tradingSettingsProvider->getMinTrailingStepForStop();
        int minTrailingStepForLimit = tradingSettingsProvider->getMinTrailingStepForLimit();
        int minTrailingStepForEntryStop = tradingSettingsProvider->getMinTrailingStepForEntryStop();
        int minTrailingStepForEntryLimit = tradingSettingsProvider->getMinTrailingStepForEntryLimit();
        int maxTrailingStepForStop = tradingSettingsProvider->getMaxTrailingStepForStop();
        int maxTrailingStepForLimit = tradingSettingsProvider->getMaxTrailingStepForLimit();
        int maxTrailingStepForEntryStop = tradingSettingsProvider->getMaxTrailingStepForEntryStop();
        int maxTrailingStepForEntryLimit = tradingSettingsProvider->getMaxTrailingStepForEntryLimit();
        double mmr = tradingSettingsProvider->getMMR(instrument, account);
        double mmr2, emr, lmr;
        bool threeLevelMargin = tradingSettingsProvider->getMargins(instrument, account, mmr2, emr, lmr);

        char buffer[256];
        const char* marketStatusStr = "unknown";
        switch (marketStatus)
        {
        case MarketStatusOpen:
            marketStatusStr = "Market Open";
            break;
        case MarketStatusClosed:
            marketStatusStr = "Market Close";
            break;
        }
        sprintf_s(buffer, sizeof(buffer), "  [%s (%s)] Cond.Dist: ST=%d; LT=%d; SE=%d; LE=%d", instrument, marketStatusStr,
            condDistStopForTrade, condDistLimitForTrade, condDistStopForEntryOrder, condDistLimitForEntryOrder);
        mLog->logString(buffer);
        sprintf_s(buffer, sizeof(buffer), "            Cond.Dist entry stop=%d; entry limit=%d", condDistEntryStop,
            condDistEntryLimit);
        mLog->logString(buffer);
        sprintf_s(buffer, sizeof(buffer), "            Quantity: Min=%d; Max=%d. Base unit size=%d; MMR=%f", minQuantity,
            maxQuantity, baseUnitSize, mmr);
        mLog->logString(buffer);
        if (threeLevelMargin)
            sprintf_s(buffer, sizeof(buffer), "            Three level margin: MMR=%f; EMR=%f; LMR=%f", mmr2, emr, lmr);
        else
            sprintf_s(buffer, sizeof(buffer), "            Single level margin: MMR=%f; EMR=%f; LMR=%f", mmr2, emr, lmr);
        mLog->logString(buffer);
        sprintf_s(buffer, sizeof(buffer), "            Trailing stop: Stop=%d-%d; Limit=%d-%d", minTrailingStepForStop,
            maxTrailingStepForStop, minTrailingStepForLimit, maxTrailingStepForLimit);
        mLog->logString(buffer);
        sprintf_s(buffer, sizeof(buffer), "            Trailing stop: Entry Stop=%d-%d; Entry Limit=%d-%d",
            minTrailingStepForEntryStop, maxTrailingStepForEntryStop, minTrailingStepForEntryLimit, maxTrailingStepForEntryLimit);
        mLog->logString(buffer);
    }

    mLog->dumpLog();
}

